src/ILCompiler.RyuJit/src/JitInterface/CorInfoImpl.RyuJit.cs

```diff
@@ -867,5 +887,61 @@ private ISymbolNode GetGenericLookupHelper(CORINFO_RUNTIME_LOOKUP_KIND runtimeLo
             return _compilation.NodeFactory.ReadyToRunHelperFromDictionaryLookup(helperId, helperArgument, MethodBeingCompiled);
         }
 
+        private CorInfoHelpFunc getCastingHelper(ref CORINFO_RESOLVED_TOKEN pResolvedToken, bool fThrowing)
+        {
+            TypeDesc type = HandleToObject(pResolvedToken.hClass);
+
+            CorInfoHelpFunc helper;
+
+            if (type.IsCanonicalDefinitionType(CanonicalFormKind.Any))
+            {
+                // In shared code just use the catch-all helper for type variables, as the same
+                // code may be used for interface/array/class instantiations
+                //
+                // We may be able to take advantage of constraints to select a specialized helper.
+                // This optimizations does not seem to be warranted at the moment.
+                helper = CorInfoHelpFunc.CORINFO_HELP_ISINSTANCEOFANY;
+            }
+            else if (type.IsInterface)
+            {
+                // If it is an interface, use the fast interface helper
+                helper = CorInfoHelpFunc.CORINFO_HELP_ISINSTANCEOFINTERFACE;
+            }
+            else if (type.IsArray)
+            {
+                // If it is an array, use the fast array helper
+                helper = CorInfoHelpFunc.CORINFO_HELP_ISINSTANCEOFARRAY;
+            }
+            else if (type.IsDefType)
+            {
+                helper = CorInfoHelpFunc.CORINFO_HELP_ISINSTANCEOFCLASS;
+#if !SUPPORT_JIT
+                // When this assert is hit, we'll have to do something with the class checks in RyuJIT
+                // Frozen strings might end up failing inlined checks generated by RyuJIT for sealed classes.
+                Debug.Assert(!_compilation.NodeFactory.CompilationModuleGroup.CanHaveReferenceThroughImportTable);
+#endif
+            }
+            else
+            {
+                // Otherwise, use the slow helper
```

Comment:
Pointers/function pointers/byrefs. However useless such type checks would be.

Reply:
Does that mean that [this](https://github.com/dotnet/corert/blob/master/src/ILCompiler.Compiler/src/Compiler/JitHelper.cs#L329) is not quite right?

Reply:
Turns out `RhTypeCast_IsInstanceOf`/`CheckCast` wouldn't be able to handle this either. We can file an issue on this to clean it up, but I don't see any scenario where this would be necessary. Maybe to support an obfuscator.